main.go

```diff
@@ -0,0 +1,138 @@
+package main
+
+import (
+    "errors"
+    "fmt"
+)
+
+type Node struct {
+    Data   int
+    Height int
+    Left   *Node
+    Right  *Node
+}
+
+type Tree struct {
+    Root *Node
+}
+
+func height(n *Node) int {
+    if n == nil {
+        return 0
+    }
+    return n.Height
+}
+
+func max(a, b int) int {
+    if a > b {
+        return a
+    }
+    return b
+}
+
+func abs(a int) int {
+    if a > 0 {
+        return a
+    }
+    return -a
+}
+
+func balance_factor(n *Node) int {
+    if n == nil {
+        return 0
+    }
+    return height(n.Left) - height(n.Right)
+}
+
+func (self *Node) insert(data int) *Node {
+    if self == nil {
```

Comment:
Can you imagine a condition where this code will be executed? 

 if the object is null, you will never get to function call.

It should be ```self.Root ===  nil```.